Modules/Scripted/Endoscopy/Endoscopy.py

```diff
@@ -834,11 +722,186 @@ def interpolateOrientations(self):
+        # delete the entries associated with this resampledCurvePointIndexToDelete even in the case that its
+        # distance has changed.
+
+        # We must freeze the cameraOrientations.keys() generator at the start (by converting it to a list) because we
+        # delete from the cameraOrientations Python dict in this loop.
+        for distanceAlongResampledCurve in list(cameraOrientations.keys()):
+            resampledCurvePointIndex = EndoscopyLogic.indexOfControlPointForDistanceAlongCurve(
+                self.resampledCurve, distanceAlongResampledCurve,
+            )
+            if resampledCurvePointIndex == resampledCurvePointIndexToDelete:
+                del cameraOrientations[distanceAlongResampledCurve]
+                break
+
+        EndoscopyLogic.setInputCurveCameraOrientations(self.inputCurve, cameraOrientations)
+        self.interpolateOrientationsForControlPoints(cameraOrientations)
+
+        return cameraOrientations
+
+    def updateCameraFromOrientationAtIndex(self, resampledCurvePointIndex):
+        """Apply the resampledCurvePointIndex-th step in the path to the camera"""
+
+        if (
+            self.resampledCurve is None
+            or not 0 <= resampledCurvePointIndex < self.getNumberOfControlPoints()
+        ):
+            return
+
+        cameraPosition = np.zeros((3,))
+        self.resampledCurve.GetNthControlPointPositionWorld(resampledCurvePointIndex, cameraPosition)
+
+        focalPoint = np.zeros((3,))
+        self.resampledCurve.GetNthControlPointPositionWorld(resampledCurvePointIndex + 1, focalPoint)
+
+        worldOrientation = np.zeros((4,))
+        self.resampledCurve.GetNthControlPointOrientation(resampledCurvePointIndex, worldOrientation)
+
+        worldMatrix3x3 = EndoscopyLogic.orientationToMatrix3x3(worldOrientation)
+        worldMatrix4x4 = EndoscopyLogic.buildCameraMatrix4x4(cameraPosition, worldMatrix3x3)
+
+        adjustedFocalPoint = cameraPosition + worldMatrix3x3[:, 2] * (
+            np.linalg.norm(focalPoint - cameraPosition) / np.linalg.norm(worldMatrix3x3[:, 2])
+        )
+
+        # Update the camera
+        cameraNode = EndoscopyLogic.getCameraFromInputCurve(self.inputCurve)
+        if cameraNode:
+            with slicer.util.NodeModify(cameraNode):
+                cameraNode.SetPosition(*cameraPosition)
+                cameraNode.SetFocalPoint(*adjustedFocalPoint)
+                cameraNode.SetViewUp(*worldMatrix3x3[:, 1])
+
+            cameraNode.ResetClippingRange()
+
+        # Update the cursor transform
+        transform = EndoscopyLogic.getTransformFromInputCurve(self.inputCurve)
+        if transform:
+            transform.SetMatrixTransformToParent(worldMatrix4x4)
+
+        return worldMatrix4x4
+
+    @staticmethod
+    def setInputCurveCamera(inputCurve, cameraNode):
+        if not inputCurve:
+            return
+        inputCurve.SetNodeReferenceID("Camera", cameraNode.GetID() if cameraNode else None)
```

Comment:
Several of these new `EndoscopyLogic` methods are static and accept an `inputCurve` parameter that, in practice, is always the `self.inputCurve` member of the `EndoscopyWidget`.  However, the `EndoscopyLogic` instance already has a `self.inputCurve` member that shadows that from `EndoscopyWidget`.  Perhaps it would be less dangerous to make these methods be non-static and use the `EndoscopyLogic` member variable, rather than accepting a potentially arbitrary `inputCurve` with a static method.